README.md
# fijkplayer (Video player plugin for Flutter) Flutter 媒体播放器

✨ **[手把手带你写 Flutter 系统音量插件](https://www.yuque.com/befovy/share/flutter_volume)**  ✨  **[Flutter 多版本管理工具 fvm](https://github.com/befovy/fvm)** ✨

[![HitCount](https://hits.dwyl.com/befovy/fijkplayer.svg)](https://hits.dwyl.com/befovy/fijkplayer) &nbsp; &nbsp;
[![pub package](https://img.shields.io/pub/v/fijkplayer.svg)](https://pub.dartlang.org/packages/fijkplayer) &nbsp; &nbsp;
[![Action Status](https://github.com/befovy/fijkplayer/workflows/Flutter/badge.svg?branch=master)](https://github.com/befovy/fijkplayer/actions) &nbsp; &nbsp;


A Flutter media player plugin for iOS and android based on [ijkplayer](https://github.com/befovy/ijkplayer)

您的支持是我们开发的动力。 欢迎Star，欢迎PR~。
[Feedback welcome](https://github.com/befovy/fijkplayer/issues) and
[Pull Requests](https://github.com/befovy/fijkplayer/pulls) are most welcome!

## Documentation 文档

* Development Documentation https://fijkplayer.befovy.com/docs/en/ quick start、guide、and concepts about fijkplayer 
* 开发文档  https://fijkplayer.befovy.com/docs/zh/ 包含快速开始、使用指南、fijkplayer 中的概念理解
* dart api https://pub.dev/documentation/fijkplayer/ detail API and argument explaination
* Release Notes https://github.com/befovy/fijkplayer/releases and [CHANGELOG.md](./CHANGELOG.md)
* FAQ https://fijkplayer.befovy.com/docs/zh/faq.html

## Installation 安装

Add `fijkplayer` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/). 

[![pub package](https://img.shields.io/pub/v/fijkplayer.svg)](https://pub.dartlang.org/packages/fijkplayer)

```yaml
dependencies:
  fijkplayer: ^{{latest version}}
```

Replace `{{latest version}}` with the version number in badge above.

Use git branch which not published to pub.
```yaml
dependencies:
  fijkplayer:
    git:
      url: https://github.com/befovy/fijkplayer.git
      ref: develop # can be replaced to branch or tag name
```

## Example 示例

```dart
import 'package:fijkplayer/fijkplayer.dart';
import 'package:flutter/material.dart';

class VideoScreen extends StatefulWidget {
  final String url;

  VideoScreen({@required this.url});

  @override
  _VideoScreenState createState() => _VideoScreenState();
}

class _VideoScreenState extends State<VideoScreen> {
  final FijkPlayer player = FijkPlayer();

  _VideoScreenState();

  @override
  void initState() {
    super.initState();
    player.setDataSource(widget.url, autoPlay: true);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(title: Text("Fijkplayer Example")),
        body: Container(
          alignment: Alignment.center,
          child: FijkView(
            player: player,
          ),
        ));
  }

  @override
  void dispose() {
    super.dispose();
    player.release();
  }
}

```

## Contributors 贡献者 ✨

Thanks goes to [these wonderful people](./CONTRIBUTORS.md) ([emoji key](https://allcontributors.org/docs/en/emoji-key))

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome

## iOS Warning 警告

Warning: The fijkplayer video player plugin is not functional on iOS simulators. An iOS device must be used during development/testing. For more details, please refer to this [issue](https://github.com/flutter/flutter/issues/14647).


## Join Ding Talk Group 加入钉钉群

<div>
  <table>
    <thead><tr>
      <th>加入钉钉群</th>
      <th>微信赞赏码</th>
      <th>支付宝</th>
    </tr></thead>
    <tbody><tr>
      <td>
        <img width="200" height="200" src="https://cdn.jsdelivr.net/gh/befovy/fijkplayer@master/docs/images/dingtalk.jpg" alt="加入钉钉群" />
      </td>
      <td>
        <img width="200" height="200" src="https://cdn.jsdelivr.net/gh/befovy/images@master/assets/wechat-qr-code.jpeg" alt="微信赞赏码" />
      </td>
      <td>
        <img width="200" height="200" src="https://cdn.jsdelivr.net/gh/befovy/images@master/assets/alipay-qr-code.jpeg" alt="支付宝二维码" />
      </td>
    </tr></tbody>
  </table>
</div>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

---
## [0.11.0](https://github.com/befovy/fijkplayer/compare/v0.10.1...v0.11.0) (2023-07-23)
* migrate to sdk: ">=2.14.0 <4.0.0"

---
## [0.10.1](https://github.com/befovy/fijkplayer/compare/v0.10.0...v0.10.1) (2021-11-30)

* fix fix mov seek null pointer crash (https://github.com/befovy/ijkplayer/pull/71) (https://github.com/befovy/FFmpeg/commit/72cfdff6)

---
## [0.10.0](https://github.com/befovy/fijkplayer/compare/v0.9.0...v0.10.0) (2021-07-11)

* download android symbols from github release.

---
## [0.9.0](https://github.com/befovy/fijkplayer/compare/v0.8.8...v0.9.0) (2021-05-29)

* null safety support
---
## [0.8.8](https://github.com/befovy/fijkplayer/compare/v0.8.7...v0.8.8) (2021-05-15)

* fix ios setLoop never return ([2476a3a](https://github.com/befovy/fijkplayer/commit/2476a3ac3d4ed94e315b11061dbfb646fa406046)), closes [#396](https://github.com/befovy/fijkplayer/issues/396) and [#366](https://github.com/befovy/fijkplayer/issues/366)

* upgrade iOS pod dependency, resue github release, fix jcenter unavaiable ([86deb1f](https://github.com/befovy/fijkplayer/commit/86deb1f77716ff54cd9df514b292dc9c1e2de5c5))

---
## [0.8.7](https://github.com/befovy/fijkplayer/compare/v0.8.6...v0.8.7) (2020-07-11)

* upgrade iOS pod dependency, speed up pod install

---
## [0.8.6](https://github.com/befovy/fijkplayer/compare/v0.8.5...v0.8.6) (2020-07-06)

* fix android context null pointer error ([4cdf670](https://github.com/befovy/fijkplayer/commit/4cdf6705a1c49eb075f3b195c4052d2281de6877)), closes [#272](https://github.com/befovy/fijkplayer/issues/272)

---

## [0.8.5](https://github.com/befovy/fijkplayer/compare/v0.8.4...v0.8.5) (2020-07-04)

* fix slider pos after seek and fullscreen toggle ([354b436](https://github.com/befovy/fijkplayer/commit/354b436fe5f3c4e73b93e985822cacfb1633b159)), closes [#261](https://github.com/befovy/fijkplayer/issues/261)
* enable iOS bitcode ([a0f1ad2](https://github.com/befovy/fijkplayer/commit/a0f1ad2e2e10a2bdb39d63bfabe4175793ccebc7))

---
## [0.8.4](https://github.com/befovy/fijkplayer/compare/v0.8.3...v0.8.4) (2020-05-16)

* fix initial volume in iOS device is zero ([fc6a60d](https://github.com/befovy/fijkplayer/commit/fc6a60d68f59730390fa7668933f28e270fc1389))
* fix pos update roll back when seeking ([064f062](https://github.com/befovy/fijkplayer/commit/064f062ab42dfd8c1831507861ff5973fca39ad3))

* add fsFit in demo and docs ([4608acb](https://github.com/befovy/fijkplayer/commit/4608acbc295ce78425cec98d65849140e1a390f6))
* lazy load android native libriries, fixes [#234](https://github.com/befovy/fijkplayer/issues/234) ([3788cfe](https://github.com/befovy/fijkplayer/commit/3788cfec1f33e590d20aff5801633185e529044d))
* fix spell error go package name ([40446af](https://github.com/befovy/fijkplayer/commit/40446afa25aefb0f9bed293ee77784277d2a3cae))

---
## [0.8.3](https://github.com/befovy/fijkplayer/compare/v0.8.2...v0.8.3) (2020-05-10)

* update ijkplayer android 0.7.4, iOS 0.7.3
* fix #226 #225 #212

---
## [0.8.2](https://github.com/befovy/fijkplayer/compare/v0.8.1...v0.8.2) (2020-05-05)

* (desktop) add global init for go-flutter, remove path_provider in demo ([7040f25](https://github.com/befovy/fijkplayer/commit/7040f25c3b4bea0de881a54f79f09cd4502fcd80))
* (android) fix color error upgrade android ijkplayer to 0.7.2 ([ae9590e](https://github.com/befovy/fijkplayer/commit/ae9590ebf7ec30141c27529fb68dd0ef8d4a8d7a))

---
## [0.8.1](https://github.com/befovy/fijkplayer/compare/v0.8.0...v0.8.1) (2020-05-04)

* Uint8List not fount ([54481ef](https://github.com/befovy/fijkplayer/commit/54481efd2167a9ac94036a0b6d7a52ee704d8296))

---

## [0.8.0](https://github.com/befovy/fijkplayer/compare/v0.7.3...v0.8.0) (2020-05-04)

* add takeSnapShot api ([c4e37eb](https://github.com/befovy/fijkplayer/commit/c4e37ebf078938e8d4b2514db87a9969e0291948))
* fix demo github action checks ([588d943](https://github.com/befovy/fijkplayer/commit/588d94355e2be0740ce0e17f912059dd64833c50))

---

## [0.7.3](https://github.com/befovy/fijkplayer/compare/v0.7.2...v0.7.3) (2020-05-01)

* don't require audio session when load plugin ([839ac6a](https://github.com/befovy/fijkplayer/commit/839ac6a74448edd31458fffcd02502f423597fae)), closes [#219](https://github.com/befovy/fijkplayer/issues/219)

---

## [0.7.2](https://github.com/befovy/fijkplayer/compare/v0.7.1...v0.7.2) (2020-04-19)


* iOS orientation changed after enter and then leave background ([d57f514](https://github.com/befovy/fijkplayer/commit/d57f5148c633bef887c8197258fc94da1f7bc2f0)), closes [#209](https://github.com/befovy/fijkplayer/issues/209)

---

## [0.7.1](https://github.com/befovy/fijkplayer/compare/v0.7.0...v0.7.1) (2020-04-11)

* update ijkplayer to 0.6.0

---
## [0.7.0](https://github.com/befovy/fijkplayer/compare/v0.6.3...v0.7.0) (2020-03-22)

* feat: migrate to flutter 1.12 android new API ([ee02c09](https://github.com/befovy/fijkplayer/commit/ee02c0912910403afc46ded9e780cd05456f60cd))
* fix: api changed, build failed using flutter beta version ([0aa1e9f](https://github.com/befovy/fijkplayer/commit/0aa1e9ffbaa18c96c285a49f55890f05343c7e24))

---
## [0.6.3](https://github.com/befovy/fijkplayer/compare/v0.6.1...v0.6.3) (2020-03-21)

* change pod library name, solve name conflict

## [0.6.2](https://github.com/befovy/fijkplayer/compare/v0.6.1...v0.6.2) (2020-03-08)

* fix error int not double  ([#195](https://github.com/befovy/fijkplayer/issues/195)) ([643a3c2](https://github.com/befovy/fijkplayer/commit/643a3c2848a9f6694a367c6db2ba78c527c607a7))

## [0.6.1](https://github.com/befovy/fijkplayer/compare/v0.6.0...v0.6.1) (2020-03-08)

* docs: add fijkPanel2Builder docs

## [0.6.0](https://github.com/befovy/fijkplayer/compare/v0.5.2...v0.6.0) (2020-03-08)


go/go-player.go
//MIT License
//
//Copyright (c) [2019] [Befovy]
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

package fijkplayer

import (
	"fmt"
	"github.com/go-flutter-desktop/go-flutter"
	"github.com/go-flutter-desktop/go-flutter/plugin"
	"net/url"
	"os"
	"path/filepath"
	"sync/atomic"
)

var atomicId int32 = 0

const (
	idle           = 0
	initialized    = 1
	asyncPreparing = 2
	prepared       = 3
	started        = 4
	paused         = 5
	completed      = 6
	stopped        = 7
	errored        = 8
	end            = 9
)

const (
	FFP_OPT_CATEGORY_FORMAT = 1
	FFP_OPT_CATEGORY_CODEC  = 2
	FFP_OPT_CATEGORY_SWS    = 3
	FFP_OPT_CATEGORY_PLAYER = 4
	FFP_OPT_CATEGORY_SWR    = 5
)
const (
	IJKMPET_FLUSH                  = 0
	IJKMPET_ERROR                  = 100
	IJKMPET_PREPARED               = 200
	IJKMPET_COMPLETED              = 300
	IJKMPET_VIDEO_SIZE_CHANGED     = 400
	IJKMPET_SAR_CHANGED            = 401
	IJKMPET_VIDEO_RENDERING_START  = 402
	IJKMPET_AUDIO_RENDERING_START  = 403
	IJKMPET_VIDEO_ROTATION_CHANGED = 404
	IJKMPET_BUFFERING_START        = 500
	IJKMPET_BUFFERING_END          = 501
	IJKMPET_BUFFERING_UPDATE       = 502
	IJKMPET_POSITION_UPDATE		   = 510
	IJKMPET_PLAYBACK_STATE_CHANGED = 700
)

type FijkPlayer struct {
	id     int32
	state  int32
	rotate int32
	width  int32
	height int32

	ijk *ijkplayer

	hostOptions   *hostOptions
	lastBuffer    *flutter.PixelBuffer
	pixels        *flutter.PixelBuffer
	texture       flutter.Texture
	texRegistry   *flutter.TextureRegistry
	methodChannel *plugin.MethodChannel
	eventChannel  *plugin.EventChannel
	sink          *queueEventSink
}

// OnListen handles a request to set up an event stream.
func (f *FijkPlayer) OnListen(arguments interface{}, sink *plugin.EventSink) {
	f.sink.setSink(sink)
}

// OnCancel handles a request to tear down the most recently created event
// stream.
func (f *FijkPlayer) OnCancel(arguments interface{}) {
	f.sink.setSink(nil)
}

func (f *FijkPlayer) initPlayer(messenger plugin.BinaryMessenger, tex *flutter.TextureRegistry) {

	f.id = atomic.AddInt32(&atomicId, 1)
	f.state = idle
	f.texRegistry = tex
	f.texture.ID = -1
	f.sink = &queueEventSink{}

	f.hostOptions = newHostOptions()
	f.ijk = newIjkPlayer()
	f.ijk.addEventListener(f.eventListener)
	f.ijk.setOption(FFP_OPT_CATEGORY_PLAYER, "overlay-format", "fcc-rgba")
	f.ijk.setIntOption(FFP_OPT_CATEGORY_PLAYER,"enable-position-notify", 1)

	f.methodChannel = plugin.NewMethodChannel(messenger,
		fmt.Sprintf("befovy.com/fijkplayer/%d", f.id),
		plugin.StandardMethodCodec{})
	f.methodChannel.HandleFunc("setupSurface", f.handleSetupSurface)
	f.methodChannel.HandleFunc("setOption", f.handleSetOption)
	f.methodChannel.HandleFunc("applyOptions", f.handleApplyOptions)
	f.methodChannel.HandleFunc("setDataSource", f.handleSetDataSource)
	f.methodChannel.HandleFunc("prepareAsync", f.handlePrepareAsync)
	f.methodChannel.HandleFunc("start", f.handleStart)
	f.methodChannel.HandleFunc("pause", f.handlePause)
	f.methodChannel.HandleFunc("stop", f.handleStop)
	f.methodChannel.HandleFunc("reset", f.handleReset)
	f.methodChannel.HandleFunc("getCurrentPosition", f.handleGetCurrentPosition)
	f.methodChannel.HandleFunc("setVolume", f.handleSetVolume)
	f.methodChannel.HandleFunc("seekTo", f.handleSeekTo)
	f.methodChannel.HandleFunc("setLoop", f.handleSetLoop)
	f.methodChannel.HandleFunc("setSpeed", f.handleSetSpeed)
	f.methodChannel.HandleFunc("snapshot", f.handleSnapshot)

	f.eventChannel = plugin.NewEventChannel(messenger,
		fmt.Sprintf("befovy.com/fijkplayer/event/%d", f.id),
		plugin.StandardMethodCodec{})
	f.eventChannel.Handle(f)
}

func (f *FijkPlayer) getId() int32 {
	return f.id
}

func (f *FijkPlayer) eventListener(what int, arg1, arg2 int32, extra string) {
	f.handleEvent(what, arg1, arg2, extra)
}

func (f *FijkPlayer) pixelReceived(buffer *flutter.PixelBuffer) {
	//pix := make([]uint8, buffer.Width * buffer.Height *4)
	//fmt.Println(buffer)
	//spew.Dump(buffer)
	//copy(pix, buffer.Pix)
	//f.pixels = &flutter.PixelBuffer{Height: buffer.Height,
	//	Width:buffer.Width,Pix:pix}
	f.pixels = buffer
	err := f.texture.FrameAvailable()
	if err != nil {
		fmt.Println("pixelReceived", err)
	}
}

func (f *FijkPlayer) handleSetupSurface(arguments interface{}) (reply interface{}, err error) {
	println("setup surface", f.texRegistry)
	if f.texRegistry != nil && f.texture.ID < 0{
		f.texture = f.texRegistry.NewTexture()
		err := f.texture.Register(f.textureHanler)
		if err != nil {
			fmt.Println(err.Error())
		}
		fmt.Printf("setupsurface tid:%d\n", f.texture.ID)
		f.ijk.setPixelCallback(f.pixelReceived)
	}
	return f.texture.ID, nil
}

func (f *FijkPlayer) handleSetOption(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	cat := -1
	key := ""
	if category, ok := args["cat"]; ok {
		cat = numInt(category, -1)
	}
	if keykey, ok := args["key"]; ok {
		key = keykey.(string)
	}
	if f.ijk != nil && cat >= 0 && len(key) > 0 {
		if intValue, exist := args["long"]; exist {
			if cat != 0 {
				f.ijk.setIntOption(cat, key, numInt64(intValue, 0))
			} else {
				f.hostOptions.addIntOption(key, numInt64(intValue, 0))
			}
		} else if strValue, exist := args["str"]; exist {
			if cat != 0 {
				f.ijk.setOption(cat, key, strValue.(string))
			} else {
				f.hostOptions.addStrOption(key, strValue.(string))
			}
		}
	}
	return nil, nil
}

func (f *FijkPlayer) handleApplyOptions(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	for o, option := range args {
		cat, oOk := o.(int)
		optionMap, optionMapOk := option.(map[interface{}]interface{})
		if oOk && optionMapOk && f.ijk != nil {
			for k, v := range optionMap {
				key, kOk := k.(string)
				intValue, intValueOk := v.(int64)
				strValue, strValueOk := v.(string)
				if kOk && intValueOk {
					if cat != 0 {
						f.ijk.setIntOption(cat, key, intValue)
					} else {
						f.hostOptions.addIntOption(key, intValue)
					}
				} else if kOk && strValueOk {
					if cat != 0 {
						f.ijk.setOption(cat, key, strValue)
					} else {
						f.hostOptions.addStrOption(key, strValue)
					}
				}
			}
		}
	}
	return nil, nil
}

func (f *FijkPlayer) handleSetDataSource(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	source, ok := args["url"]
	if ok {
		if urlStr, urlStrOk := source.(string); urlStrOk {
			u, err := url.Parse(urlStr)
			if err == nil && u != nil && u.Scheme == "asset" {
				ex, err := os.Executable()
				if err == nil {
					exPath := filepath.Dir(ex)
					urlStr = filepath.Join(exPath, "flutter_assets", u.Path)
				}
			}
			f.ijk.setDataSource(urlStr)
		}
	}
	f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, initialized, -1, nil)
	return nil, nil
}

func (f *FijkPlayer) handlePrepareAsync(arguments interface{}) (reply interface{}, err error) {
	if f.ijk != nil {
		f.ijk.prepareAsync()
	}
	f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, asyncPreparing, -1, nil)
	return nil, nil
}
func (f *FijkPlayer) handleStart(arguments interface{}) (reply interface{}, err error) {
	if f.ijk != nil {
		f.ijk.start()
	}
	return nil, nil
}

func (f *FijkPlayer) handlePause(arguments interface{}) (reply interface{}, err error) {
	if f.ijk != nil {
		f.ijk.pause()
	}
	return nil, nil
}

func (f *FijkPlayer) handleStop(arguments interface{}) (reply interface{}, err error) {
	if f.ijk != nil {
		f.ijk.stop()
	}
	f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, stopped, -1, nil)
	return nil, nil
}

func (f *FijkPlayer) handleReset(arguments interface{}) (reply interface{}, err error) {
	if f.ijk != nil {
		f.ijk.reset()
	}
	f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, idle, -1, nil)
	return nil, nil
}

func (f *FijkPlayer) handleGetCurrentPosition(arguments interface{}) (reply interface{}, err error) {
	var pos int64 = 0
	if f.ijk != nil {
		pos = f.ijk.getCurrentPosition()
	}
	return pos, nil
}

func (f *FijkPlayer) handleSetVolume(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	vol, ok := args["volume"]
	if ok {
		volume := numFloat32(vol, 1.0)
		f.ijk.setVolume(volume, volume)
	}
	return nil, nil
}

func (f *FijkPlayer) handleSeekTo(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	msec, ok := args["msec"]
	if ok {
		if f.state == completed {
			f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, paused, -1, nil)
		}
		msecInt64 := numInt64(msec, 0)
		f.ijk.seekTo(msecInt64)
	}
	return nil, nil
}

func (f *FijkPlayer) handleSetLoop(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	loop, ok := args["loop"]
	if ok {
		loopInt := numInt(loop, 1)
		f.ijk.setLoop(loopInt)
	}
	return nil, nil
}

func (f *FijkPlayer) handleSetSpeed(arguments interface{}) (reply interface{}, err error) {
	args := arguments.(map[interface{}]interface{})
	speed, ok := args["speed"]
	if ok {
		speedFloat32 := numFloat32(speed, 1.0)
		f.ijk.setSpeed(speedFloat32)
	}
	return nil, nil
}

func (f* FijkPlayer) handleSnapshot(arguments interface{}) (reply interface{}, err error)  {
	_ = f.methodChannel.InvokeMethod("_onSnapshot", "not support")
	return nil, nil
}

func (f *FijkPlayer) textureHanler(width, height int) (bool, *flutter.PixelBuffer) {
	// t := time.Now()
	// fmt.Println(t.Unix(), "textureHanler", width, height)
	//return true, &flutter.PixelBuffer{Width:16, Height:16,Pix: pix}
	return true, f.pixels
}

func (f *FijkPlayer) release() {

	f.handleEvent(IJKMPET_PLAYBACK_STATE_CHANGED, end, f.state, nil)

	if f.ijk != nil {
		f.ijk.shutdown()
		f.ijk = nil
	}
	if f.texRegistry != nil {
		f.texRegistry = nil
		fmt.Printf("texture %d\n", f.texture.ID)
		if f.texture.ID >= 0 {
			err := f.texture.UnRegister()
			if err != nil {
				fmt.Printf("unRegister %s\n", err.Error())
			}
			f.texture.ID = -1
		}
	}

	f.methodChannel.ClearAllHandle()
	f.methodChannel.CatchAllHandleFunc(nil)
	f.methodChannel = nil
	f.sink.setSink(nil)
	f.eventChannel.Handle(nil)
	f.eventChannel = nil
}

func isPlayable(state int32) bool {
	return state == started || state == paused || state == completed || state == prepared
}

func onStateChanged(newState int32, oldState int32) {
	fpg := fplugin
	if fpg != nil {
		if newState == started && oldState != started {
			fpg.onPlayingChange(1)
		} else if newState != started && oldState == started {
			fpg.onPlayingChange(-1)
		}

		if isPlayable(newState) && !isPlayable(oldState) {
			fpg.onPlayableChange(1)
		} else if !isPlayable(newState) && isPlayable(oldState) {
			fpg.onPlayableChange(-1)
		}
	}
}

func (f *FijkPlayer) handleEvent(what int, arg1, arg2 int32, extra interface{}) {
	event := make(map[interface{}]interface{})

	switch what {
	case IJKMPET_PREPARED:
		event["event"] = "prepared"
		event["duration"] = f.ijk.getDuration()
		f.sink.success(event)
		break
	case IJKMPET_PLAYBACK_STATE_CHANGED:
		f.state = arg1
		event["event"] = "state_change"
		event["new"] = arg1
		event["old"] = arg2
		f.sink.success(event)
		onStateChanged(arg1, arg2)
		break
	case IJKMPET_VIDEO_RENDERING_START,
		IJKMPET_AUDIO_RENDERING_START:
		event["event"] = "rendering_start"
		if what == IJKMPET_VIDEO_RENDERING_START {
			event["type"] = "video"
		} else {
			event["type"] = "audio"
		}
		f.sink.success(event)
		break
	case IJKMPET_BUFFERING_START,
		IJKMPET_BUFFERING_END:
		event["event"] = "freeze"
		event["value"] = what == IJKMPET_BUFFERING_START
		f.sink.success(event)
		break
	case IJKMPET_BUFFERING_UPDATE:
		event["event"] = "buffering"
		event["head"] = arg1
		event["percent"] = arg2
		f.sink.success(event)
		break
	case IJKMPET_POSITION_UPDATE:
		event["event"] = "pos";
		event["pos"] = arg1;
		f.sink.success(event);
		break;
	case IJKMPET_VIDEO_ROTATION_CHANGED:
		event["event"] = "rotate"
		event["degree"] = arg1
		f.sink.success(event)
		f.rotate = arg1
		if f.height > 0 && f.width > 0 {
			f.handleEvent(IJKMPET_VIDEO_SIZE_CHANGED, f.width, f.height, nil)
		}
		break
	case IJKMPET_VIDEO_SIZE_CHANGED:
		event["event"] = "size_changed"
		if f.rotate == 0 || f.rotate == 90 {
			event["width"] = arg1
			event["height"] = arg2
			f.sink.success(event)
		} else if f.rotate == 90 || f.rotate == 270 {
			event["width"] = arg2
			event["height"] = arg1
			f.sink.success(event)
		}
		f.width = arg1
		f.height = arg2
		break
	case IJKMPET_ERROR:
		str := ""
		if s, ok := extra.(string); ok {
			str = s
		}
		f.sink.onError(fmt.Sprintf("%d", arg1), str, arg2)
		break
	default:
		break
	}
}
